### PR TITLE
Refactor/input

### PR DESCRIPTION
**Pull request by wjian23**

### CODE REVIEW

File: widgets/src/layout/ignore_layout.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
+use ribir_core::prelude::*;
+
+/// When the IgnoreLayoutChanged wrap the widget, the subtree's inner change
+/// will not notify to the parent, it mean's that parent will not do the
+/// re-layout. it's
```
Comment (sologeek) [inline widgets/src/layout/ignore_layout.rs:5]: This comment `it's` next what? Or unnecessary?

**Comment by sologeek** [comment]
`input.rs` file `struct TextArea` field `multi_line` is unnecessary, can remove it.

File: widgets/src/layout/ignore_layout.rs (hunk 1 above)
Comment (wjian23) [reply widgets/src/layout/ignore_layout.rs:5]: unnecessary

File: widgets/src/layout/ignore_layout.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+#[derive(SingleChild, Query, Declare)]
+pub struct IgnoreLayoutChanged {}
```
Comment (M-Adoo) [inline widgets/src/layout/ignore_layout.rs:7]: `SizedIndependently` is a better name?

File: widgets/src/input.rs [hunk 3]
```diff
@@ -197,3 +393,66 @@ fn glyph_width(font_size: FontSize) -> f32 {
 }
```
Comment (M-Adoo) [inline widgets/src/input.rs:393]: same with `font_size.into_pixel().value()`

File: widgets/src/input.rs [hunk 4]
```diff
@@ -197,3 +393,66 @@ fn glyph_width(font_size: FontSize) -> f32 {
...
 
 fn line_height(line_height: Em) -> f32 { FontSize::Em(line_height).into_pixel().value() }
```
Comment (M-Adoo) [inline widgets/src/input.rs:395]: Pixel::from(line_height).value()

File: widgets/src/input/text_selectable.rs [hunk 5]
```diff
@@ -1,121 +1,228 @@
...
+pub(crate) fn bind_point_listener<T: SelectableText>(
+  this: impl StateWriter<Value = T>,
+  host: Widget,
+  text: Reader<Text>,
+  layout_box: Reader<LayoutBox>,
+) -> impl WidgetBuilder {
+  fn_widget! {
+    let layout1 = layout_box.clone_reader();
+    let layout2 = layout_box.clone_reader();
+    let layout3 = layout_box.clone_reader();
```
Comment (M-Adoo) [inline widgets/src/input/text_selectable.rs:78]: hint capture writer in its closure.

File: widgets/src/input/text_selectable.rs [hunk 6]
```diff
@@ -1,121 +1,228 @@
...
+
+  fn select_range(&self) -> Range<usize>;
+
+  fn text(&self) -> &CowArc<str>;
+
+  fn caret(&self) -> CaretState;
+
+  fn set_caret(&mut self, caret: CaretState);
+
+  fn select_text_rect(&self, text: &Text, text_size: Size) -> Vec<Rect> {
```
Comment (M-Adoo) [inline widgets/src/input/text_selectable.rs:34]: should not pass a `Text` widget, maybe you can provide information by methods like `SelectableText::text`, for example: `TextSelectable::text_style`.

File: widgets/src/layout.rs [hunk 7]
```diff
@@ -33,3 +33,5 @@ mod stack;
 pub use stack::*;
 pub mod constrained_box;
 pub use constrained_box::ConstrainedBox;
+pub mod ignore_layout;
```
Comment (M-Adoo) [inline widgets/src/layout.rs:36]: rename mod & file name

**Review by sologeek — APPROVED** [review]

**Review by M-Adoo — APPROVED** [review]

**Pull request merged**